src/engine.rs

```diff
@@ -219,215 +191,40 @@ impl FileEngineInner {
-                    Some(file_num) => file_num,
-                    None => continue,
-                };
-
-                // Has entries left behind too far, this happens when
-                // some followers left behind for a long time.
-                if min_file_num < gc_file_num {
-                    info!(
-                        "region {}'s some followers left behind too far, \
-                         need force compaction",
-                        memtable.region_id()
-                    );
-                    regions.insert(memtable.region_id());
-                }
-            }
-        }
-        NEED_COMPACT_REGIONS_HISTOGRAM.observe(regions.len() as f64);
-
-        regions
-    }
-
-    #[allow(dead_code)]
-    fn evict_old_from_cache(&self) {
-        let inactive_file_num = self
-            .pipe_log
-            .files_before(self.cfg.cache_size_limit.0 as usize);
-
-        if inactive_file_num == 0 {
-            return;
-        }
-
-        for slot in 0..SLOTS_COUNT {
-            let mut memtables = self.memtables[slot].write().unwrap();
-            for memtable in memtables.values_mut() {
-                memtable.evict_old_from_cache(inactive_file_num);
-            }
-        }
-    }
-
-    #[allow(dead_code)]
     fn purge_expired_files(&self) -> Result<()> {
         let mut min_file_num = u64::MAX;
         for memtables in &self.memtables {
             let memtables = memtables.read().unwrap();
-            let file_num = memtables.values().fold(u64::MAX, |min, x| {
-                cmp::min(min, x.min_file_num().map_or(u64::MAX, |num| num))
-            });
-            if file_num < min_file_num {
-                min_file_num = file_num;
+            for memtable in memtables.values() {
+                if let Some(file_num) = memtable.min_file_num() {
+                    min_file_num = cmp::min(min_file_num, file_num);
+                }
             }
         }
-
         self.pipe_log.purge_to(min_file_num)
     }
 
-    fn compact_to(&self, region_id: u64, index: u64) -> u64 {
-        let mut memtables = self.memtables[region_id as usize % SLOTS_COUNT]
-            .write()
+    fn first_index(&self, region_id: u64) -> Option<u64> {
+        let memtables = self.memtables[region_id as usize % SLOTS_COUNT]
+            .read()
             .unwrap();
-        if let Some(memtable) = memtables.get_mut(&region_id) {
-            return memtable.compact_to(index);
+        if let Some(memtable) = memtables.get(&region_id) {
+            return memtable.first_index();
         }
-        0
+        None
+    }
+
+    fn compact_to(&self, region_id: u64, index: u64) -> u64 {
+        let first_index = match self.first_index(region_id) {
+            Some(index) => index,
+            None => return 0,
+        };
+
+        let log_batch = LogBatch::new();
+        log_batch.add_command(Command::Compact { region_id, index });
+        self.write(log_batch, true).map(|_| ()).unwrap();
```

Comment:
I prefer buffer write for compact command, this will reduce the consumption of fsync call.